### PR TITLE
remove redundent "<>" for ty::Slice with reference type

### LINKED ISSUE #103271: check_for_deref_method suggestion introduces a parse error
Given the following code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=698ee23815c463c0aa5018a839b8a43f)):

```rust
fn main() {
    let length = <&[u32]>::len;
    assert_eq!(length(&[1,3]), 2);
}
```

The current output is:

```
error[E0599]: the function or associated item `len` exists for reference `&[u32]`, but its trait bounds were not satisfied
 --> src/main.rs:2:28
  |
2 |     let length = <&[u32]>::len;
  |                            ^^^ function or associated item cannot be called on `&[u32]` due to unsatisfied trait bounds
  |
  = note: the following trait bounds were not satisfied:
          `&[u32]: ExactSizeIterator`
          which is required by `&mut &[u32]: ExactSizeIterator`
help: the function `len` is implemented on `[u32]`
  |
2 |     let length = <<[u32]>>::len;
  |                   ~~~~~~~
```

This suggestion is on the right track, but unfortunately it introduces a parse error. It should be:

```
help: the function `len` is implemented on `[u32]`
  |
2 |     let length = <[u32]>::len;
  |                   ~~~~~
```

This suggestion is in [check_for_deref_method](https://github.com/rust-lang/rust/blob/4b8f4319954ff2642690b9e5cbe4af352d095bf6/compiler/rustc_hir_analysis/src/check/method/suggest.rs#L1830-L1884). It suggests adding `<>` around the type because the type is a slice, but I already have `<>` in my program (and the span doesn't include the existing `<>`).

### PR DESCRIPTION
**Pull request by lyming2007**
this fix #103271

### CODE REVIEW

**Review by fee1-dead — CHANGES_REQUESTED** [review]
There might be a better way. AFAICT this diagnostic is erroneous because the span it covers does not contain the angle brackets, so it prints the suggestion on `[_]` instead of `<[_]>`. I believe this can be done pretty cheaply using [`span_to_source`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_span/source_map/struct.SourceMap.html#method.span_to_source), skipping white space and enlarging the span if you see angle brackets from both sides.

**Comment by lyming2007** [comment]
> There might be a better way. AFAICT this diagnostic is erroneous because the span it covers does not contain the angle brackets, so it prints the suggestion on `[_]` instead of `<[_]>`. I believe this can be done pretty cheaply using [`span_to_source`](https://doc.rust-lang.org/beta/nightly-rustc/rustc_span/source_map/struct.SourceMap.html#method.span_to_source), skipping white space and enlarging the span if you see angle brackets from both sides.

pushed a new fix by seeing angle brackets from both sides.

File: compiler/rustc_hir_typeck/src/method/suggest.rs
Comment (fee1-dead) [inline compiler/rustc_hir_typeck/src/method/suggest.rs:1873]: I don't think it should just be `ty::Slice` as we recommend `<{deref_ty}>` to all types below. So we should stop recommending `<{deref_ty}>` no matter what `deref_ty` is if there are angle brackets already.

**Review by fee1-dead — COMMENTED** [review]
One last nit. Should be good to go after this.

File: compiler/rustc_hir_typeck/src/method/suggest.rs
Comment (lyming2007) [reply compiler/rustc_hir_typeck/src/method/suggest.rs]: yes. You're right. We should not recommend <{deref_ty}> only for ty::Slice.
And the change was pushed.
Thank you for the suggestions.

**Comment by fee1-dead** [comment]
Looks good to me, thanks for this PR!

@bors r+

**Comment by Dylan-DPC** [comment]
failed in [rollup](https://github.com/rust-lang/rust/pull/103921#issuecomment-1302038882)

@bors r-

**Comment by lyming2007** [comment]
@fee1-dead should I do something for the merging conflict? This is the first time I encounter this type of conflict

**Comment by Dylan-DPC** [comment]
What merging conflict? I don't see any conflict in your PR.

**Comment by lyming2007** [comment]
> failed in [rollup](https://github.com/rust-lang/rust/pull/103921#issuecomment-1302038882)



> failed in [rollup](https://github.com/rust-lang/rust/pull/103921#issuecomment-1302038882)

It says failed in https://github.com/rust-lang/rust/pull/103921#issuecomment-1302038882

**Comment by Dylan-DPC** [comment]
yes it means your pr failed some tests when it was rolled up. You should see the failures in the [error log](https://github.com/rust-lang-ci/rust/actions/runs/3385566442/jobs/5623869747) (after a bit of scrolilng ;P). You need to update those tests and then we can merge this again

**Comment by lyming2007** [comment]
> yes it means your pr failed some tests when it was rolled up. You should see the failures in the [error log](https://github.com/rust-lang-ci/rust/actions/runs/3385566442/jobs/5623869747) (after a bit of scrolling ;P). You need to update those tests and then we can merge this again

I changed according to that error log. But the CI didn't pass:
```
---- [ui] src/test/ui/type/issue-103271.rs stdout ----
diff of stderr:

7	   = note: the following trait bounds were not satisfied:
8	           `&[_]: ExactSizeIterator`
9	           which is required by `&mut &[_]: ExactSizeIterator`
+	   = help: items from traits can only be used if the trait is in scope
+	help: the following trait is implemented but not in scope; perhaps add a `use` for it:
+	   |
+	LL | use object::read::read_ref::ReadRef;
+	   |
10	help: the function `len` is implemented on `[_]`
11	   |
12	LL |     let length = <[_]>::len;
```

**Comment by lyming2007** [comment]
> yes it means your pr failed some tests when it was rolled up. You should see the failures in the [error log](https://github.com/rust-lang-ci/rust/actions/runs/3385566442/jobs/5623869747) (after a bit of scrolling ;P). You need to update those tests and then we can merge this again

I think my last code should fix the in the [error log](https://github.com/rust-lang-ci/rust/actions/runs/3385566442/jobs/5623869747). can you merge it again?

**Comment by lyming2007** [comment]
@Dylan-DPC any progress on it?

**Comment by Dylan-DPC** [comment]
The CI doesn't seem to pass though? let me rerun it. Also i would wait for @fee1-dead 's confirmation whether it is fine to be merged again or not

**Comment by lyming2007** [comment]
> in the [error log](https://github.com/rust-lang-ci/rust/actions/runs/3385566442/jobs/5623869747)

I fixed it according to the error of this log [error log](https://github.com/rust-lang-ci/rust/actions/runs/3385566442/jobs/5623869747)
which seems conflict with the CI in PR https://github.com/rust-lang/rust/pull/103675

**Comment by fee1-dead** [comment]
I have restarted the CI to see if this is fixed.

**Comment by lyming2007** [comment]
> I have restarted the CI to see if this is fixed.
the CI of this PR used to work.
It requires something like that:
```
---- [ui] src/test/ui/type/issue-103271.rs stdout ----
diff of stderr:

7	   = note: the following trait bounds were not satisfied:
8	           `&[_]: ExactSizeIterator`
9	           which is required by `&mut &[_]: ExactSizeIterator`
+	   = help: items from traits can only be used if the trait is in scope
+	help: the following trait is implemented but not in scope; perhaps add a `use` for it:
+	   |
+	LL | use object::read::read_ref::ReadRef;
+	   |
10	help: the function `len` is implemented on `[_]`
11	   |
12	LL |     let length = <[_]>::len;
```
but the failures in the [error log](https://github.com/rust-lang-ci/rust/actions/runs/3385566442/jobs/5623869747) requires something like that:
```
failures:

---- [ui] src/test/ui/type/issue-103271.rs stdout ----
diff of stderr:

7	   = note: the following trait bounds were not satisfied:
8	           `&[_]: ExactSizeIterator`
9	           which is required by `&mut &[_]: ExactSizeIterator`
-	   = help: items from traits can only be used if the trait is in scope
-	help: the following trait is implemented but not in scope; perhaps add a `use` for it:
-	   |
-	LL | use object::read::read_ref::ReadRef;
-	   |
15	help: the function `len` is implemented on `[_]`
16	   |
17	LL |     let length = <[_]>::len;


```
So they're conflicting with each other.  Can I remove the test case and you approve the PR? after the code is merged, I'll add the test case later. @fee1-dead

**Comment by fee1-dead** [comment]
Could you try [rebasing](https://rustc-dev-guide.rust-lang.org/git.html#keeping-things-up-to-date) your PR on top of master?

**Comment by fee1-dead** [comment]
Also to expand on my last point, the PR's CI must pass as if it fails then the test jobs will most likely fail after I approve. The failure of the rollup was `test-various`, which is only one of the multiple test jobs that we have. It could be spurious though and we could retry after CI passes.

**Comment by lyming2007** [comment]
> Also to expand on my last point, the PR's CI must pass as if it fails then the test jobs will most likely fail after I approve. The failure of the rollup was `test-various`, which is only one of the multiple test jobs that we have. It could be spurious though and we could retry after CI passes.

I've rebased the master and fixed the CI. Please advise. @fee1-dead

**Comment by fee1-dead** [comment]
Let's see if it will work this time.

@bors r+

**Comment by GuillaumeGomez** [comment]
Failed in https://github.com/rust-lang/rust/pull/104151.

@bors r-

**Comment by lyming2007** [comment]
the test in https://github.com/rust-lang-ci/rust/actions/runs/3420197276/jobs/5694702753 failed as I mentioned before:
```
failures:

---- [ui] src/test/ui/type/issue-103271.rs stdout ----
diff of stderr:

7	   = note: the following trait bounds were not satisfied:
8	           `&[_]: ExactSizeIterator`
9	           which is required by `&mut &[_]: ExactSizeIterator`
-	   = help: items from traits can only be used if the trait is in scope
-	help: the following trait is implemented but not in scope; perhaps add a `use` for it:
-	   |
-	LL | use object::read::read_ref::ReadRef;
-	   |
15	help: the function `len` is implemented on `[_]`
16	   |
17	LL |     let length = <[_]>::len;
```
but if I removed these code:
```
-	   = help: items from traits can only be used if the trait is in scope
-	help: the following trait is implemented but not in scope; perhaps add a `use` for it:
-	   |
-	LL | use object::read::read_ref::ReadRef;
-	   |
```
This CI would fail at:
```
---- [ui] src/test/ui/type/issue-103271.rs stdout ----
diff of stderr:

7	   = note: the following trait bounds were not satisfied:
8	           `&[_]: ExactSizeIterator`
9	           which is required by `&mut &[_]: ExactSizeIterator`
+	   = help: items from traits can only be used if the trait is in scope
+	help: the following trait is implemented but not in scope; perhaps add a `use` for it:
+	   |
+	LL | use object::read::read_ref::ReadRef;
+	   |
10	help: the function `len` is implemented on `[_]`
11	   |
12	LL |     let length = <[_]>::len;
```
@fee1-dead

**Comment by fee1-dead** [comment]
I asked [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/CI.20and.20test-various.20contradicting.20results), and this is #26454. The solution here is to use another function name that does not have the suggestion.

**Comment by lyming2007** [comment]
@fee1-dead should be ok now. please approve it again. :)

File: src/test/ui/type/issue-103271.rs
Comment (fee1-dead) [inline src/test/ui/type/issue-103271.rs:2]: I think you need to change the `len` to another function that does not interfere with the trait from the sysroot.

Comment (lyming2007) [reply src/test/ui/type/issue-103271.rs]: Changed to `iter`

**Pull request merged**
